### PR TITLE
Updated transaction metrics

### PR DESCRIPTION
**Pull request by debonair**
Collects metrics for async transactions

### CODE REVIEW

**Comment by debonair** [comment]
@devcritter @rcrichton please review. We decided to  make the asyncronous response be a secondary route so as not to overwrite the original response. I am concerned though about sending mongo queries over the wire, because they require the mediator to know too much about the underlying data model @bausmeier  what do you think

File: src/api/transactions.coffee [hunk 1]
```diff
@@ -220,12 +228,71 @@ exports.updateTransaction = (transactionId) ->
...
+    yield transactions.Transaction.findByIdAndUpdate transactionId, updates, () ->
+      that.body = "Transaction with ID: #{transactionId} successfully updated"
+      that.status = 'ok'
+      logger.info "User #{that.authenticated.email} updated transaction with id #{transactionId}"
+
+      ###
+      # Update transaction metrics
+      ###
+      transactions.Transaction.findById transactionId, (err, doc) ->
+        if updates['$push'].routes?
```
Comment (rcrichton) [inline src/api/transactions.coffee]: What is $push here. Is this a property someone has to set when updating or is it something different?

Comment (debonair) [inline src/api/transactions.coffee]: You have to set the push property so that you don't overwrite the other secondary routes with your update.

File: src/api/transactions.coffee [hunk 2]
```diff
@@ -220,12 +228,71 @@ exports.updateTransaction = (transactionId) ->
...
   transactionId = unescape transactionId
   updates = this.request.body
+  that = this
 
   try
-    yield transactions.Transaction.findByIdAndUpdate(transactionId, updates).exec()
-    this.body = "Transaction with ID: #{transactionId} successfully updated"
-    this.status = 'ok'
-    logger.info "User #{this.authenticated.email} updated transaction with id #{transactionId}"
+    yield transactions.Transaction.findByIdAndUpdate transactionId, updates, () ->
```
Comment (rcrichton) [inline src/api/transactions.coffee]: Why did you need to change this to a callback instead of just having it as it was before. This should work the same way.

Comment (debonair) [inline src/api/transactions.coffee]: I needed the stats log actions to happen after the transaction has happened, i.e in that callback I made

File: src/api/transactions.coffee (hunk 1 above)
Comment (rcrichton) [inline src/api/transactions.coffee]: Ah, I see so its a mongodb operator.

**Comment by rcrichton** [comment]
Hey @debonair this looks good to me. You may want to try rather use string interpolation to make this look neater but that isn't required. Ie. `sdc.timing "#{domain}.channels.#{doc.channelID}.#{route.name}.orchestrations.#{orchestrationName}.#{metric.name}", metric.value`

Otherwise, :+1:

File: src/api/transactions.coffee (hunk 2 above)
Comment (rcrichton) [inline src/api/transactions.coffee]: But the yield will wait until the promise is resolved before continuing. So I don't think you need the callback at all. It will do the same thing.

Comment (bausmeier) [inline src/api/transactions.coffee]: I am seeing some really weird behavior here with the yield and the callback. Errors updating the transaction aren't being handled and I am getting a combination of successful and failed responses and log messages.

File: src/api/transactions.coffee [hunk 3]
```diff
@@ -220,12 +228,71 @@ exports.updateTransaction = (transactionId) ->
...
+      that.body = "Transaction with ID: #{transactionId} successfully updated"
+      that.status = 'ok'
+      logger.info "User #{that.authenticated.email} updated transaction with id #{transactionId}"
+
+      ###
+      # Update transaction metrics
+      ###
+      transactions.Transaction.findById transactionId, (err, doc) ->
```
Comment (bausmeier) [inline src/api/transactions.coffee]: You should handle the err argument.

**Comment by debonair** [comment]
closes #403
